ziosql/src/main/scala/com/softwaremill/sql/ZioSqlTests.scala

```diff
@@ -0,0 +1,271 @@
+            _                          <- ZIO.logInfo(s"Big cities: \n ${bigCities.mkString("\n")}")
+            transactionCities          <- citiesTransaction
+            _                          <- ZIO.logInfo(s"All cities after transaction: \n ${transactionCities.mkString("\n")}")
+            _                          <- ZIO.logInfo(s"Rendered complex query: \n ${complexQuerySql} \n")
+            _                          <- ZIO.logInfo(s"Rendered insert query : \n ${insertSql} \n")
+        } yield ()).provideCustomLayer(driverLayer)
+        
+
+    // SIMPLE
+
+    /**
+      * select id, name, population, area, link from city where population > $limit
+      */
+    def citiesBiggerThan(people: Int) = 
+        select(cityId ++ cityName ++ population ++ area ++ link)
+            .from(city)
+            .where(population > people)
+            .to {
+                case (id, name, pop, area, link) =>
+                    City(CityId(id), name, pop, area, link)
+            }   
+
+    def executeCities(cityQuery: Read[City]) = 
+        execute(cityQuery)
+            .runCollect
+            .map(_.toList)
+
+    val result = executeCities(citiesBiggerThan(4000000))
+
+    // COMPLEX            
+
+    /**
+        SELECT ms.name, c.name, COUNT(ml.id) as line_count
+        FROM metro_line as ml
+        JOIN metro_system as ms on ml.system_id = ms.id
+        JOIN city AS c ON ms.city_id = c.id
+        GROUP BY ms.name, c.name
+        ORDER BY line_count DESC
+    */
+    val lineCount = (Count(metroLineId) as "line_count")              
+
+    val complexQuery = select(metroLineName ++ cityName ++ lineCount)
+        .from(metroLine
+            .join(metroSystem).on(metroSystemId === systemId)
+            .join(city).on(cityIdFk === cityId))
+        .groupBy(metroLineName, cityName)
+        .orderBy(Desc(lineCount))
+
+
+    // DYNAMIC
+
+    val base = select(metroLineId ++ systemId ++ metroLineName ++ stationCount ++ trackType).from(metroLine)
+
+    val minStations: Option[Int] = Some(10)
+    val maxStations: Option[Int] = None
+    val sortDesc: Boolean = true
+
+    val minStationsQuery = minStations.map(m => stationCount >= m).getOrElse(Expr.literal(true))
+    val maxStationsQuery = maxStations.map(m => trackType <= m).getOrElse(Expr.literal(true))
+
+    val ord = 
+        if (sortDesc) 
+            stationCount.desc 
+        else 
+            stationCount.asc
+
+    val whereExpr =
+        minStationsQuery && maxStationsQuery
+
+    val finalQuery = base.where(whereExpr).orderBy(ord)
+
+    // INSERT
+
+    def insertCity(id: CityId) = insertInto(city)(cityId ++ cityName ++ population ++ area ++ link)
+        .values((id.id, "London", 8982000, 1583F, Option("https://tfl.gov.uk/modes/tube/")))
+
+    // DELETE
+
+    def deleteCity(id: CityId) = 
+        deleteFrom(city)
+            .where(cityId === id.id)
+
+    // TRANSACTIONS
+
+    val i = CityId(4)
+
+    val transaction = for {
```

Comment:
this example looks different from the other "transaction" ones. Are you running a single transaction here, or are there a couple of them?

Reply:
its one transaction, I modified the example to look like the other ones